text/0000-wire-reform.md

```diff
@@ -0,0 +1,146 @@
+
+```js
+// shared adapter
+import { MyWireAdapter } from 'some-module';
+
+class Foo extends React.Component {
+    constructor(props) {
+        super(props);
+        // The wire adapter instance is bound to the host object via the callback for data
+        this.adapter = new MyWireAdapter((data) => {
+            this.setState(() => {
+                // stream of data from wire adapter to be used to update the component's state
+                return data;
+            });
+        });
+        // calling for the initial update of the config since componentDidUpdate() is not
+        // invoked for the first time, but all props are ready.
+        this.adapter.update({ x: 1, y: this.props.valueOfY });
+    }
+
+    componentDidUpdate(prevProps) {
+        if (this.props.valueOfY !== prevProps.valueOfY) {
+            // recompute the config by extracting `this.props.valueOfY`
+            this.adapter.update({ x: 1, y: this.props.valueOfY });
+        }
+    }
+
+    componentDidMount() {
+        this.adapter.connect();
+    }
+
+    componentWillUnmount() {
+        this.adapter.disconnect();
+    }
+
+    render() {
+        return (
+            <div>{this.state.valueProducedByMyWireAdapter}</div>
+        );
+    }
+}
+```
+
+# Adoption strategy
+
+Since there is a need to support callable adapters that behave differently depending on who uses that (wire adapter vs user invoking the function directly), we have added a simple mechanism to support such feature via `adapter` property member expression on the callable. This opens the door to transition existing adapters to the new form. E.g.: APEX adapters are all callable objects.
+
+Additionally, those callable objects can implement a forking logic based on the type of argument, if there is a desire to avoid the `adapter` property member expression.
+
+# How we teach this
+
+* For adapter consumers, nothing changes.
+* For adapter author, the wire protocol is now well defined and does not need registration, which means it is easier to reasoning about compared to the existing mechanism.
+* As for existing adapters based on `@lwc/wire-service`, they can remain the same until after they get refactored and simplified when possible.
+
+# Unresolved questions
+
+* In the current implementation, a wired field is `writable`, which means the component author can alter the value of the field at will. What should we do? a) throw on setter, b) do nothing on setter, c) preserve the current semantics. This is a breaking change if we do a) or b), while the current behavior is weird.
+* How context providers can be provisioned? In theory, a context provider is bound to a particular framework/system, while the context consumer is abstracted out in the wire adapter, and specific implementations per framework can provide the piping into the wire adapter protocol via the second argument in the constructor and the `adapter.context(uid, value)` mechanism. Is this sufficient?
```

Comment:
I am not sure to understand, what do you mean here?

Reply:
let's chat about this today.